content/blog/zephyr-tracing-and-profiling/index.md

```diff
@@ -0,0 +1,279 @@
+
+```c
+/* Called when data needs to be sent to network */
+int net_send_data(struct net_pkt *pkt)
+{
+    int status;
+    int ret;
+
+    SYS_PORT_TRACING_FUNC_ENTER(net, send_data, pkt);
+
+    (...)
+}
+```
+
+Instrumenting application, however, is currently not supported without manually implementing some code inside the *tracing* subsystem. For the time being, there are 3rd party tools that support such feature, and a new tool is expected to be integrated to Zephyr in the future that might as well support it [[3]].
+
+## Zephyr's profiling tool
+
+The current profiling implementation monitors the function stack in order to allow examining the software execution at given intervals. To enable it, the following configs must be defined:
+
+```text
+CONFIG_PROFILING=y
+CONFIG_PROFILING_PERF=y
+CONFIG_PROFILING_PERF_BUFFER_SIZE=2048
+CONFIG_THREAD_STACK_INFO=y
+CONFIG_SMP=n
+CONFIG_SHELL=y
+CONFIG_FRAME_POINTER=y
+```
+
+Using these directives it is possible to obtain a log that can be processed by tools and yield a composed view of what the system is executing:
+
+<a href="img/img003.webp" target="_blank">
+    {{< figure
+        default=true
+        src="img/img003.webp"
+        alt=""
+        caption="Profiling sample graph"
+    >}}
+</a>
+
+The graph above shows a timeline of the functions pushed onto the stack, including all sub-function calls. While monitoring the function stack can be useful for analyzing certain kinds of problems, we won’t discuss this feature in detail, as it offers limited functionality for now. As mentioned earlier, a new subsystem is being developed that may provide better functionality [[3]].
+
+## Percepio's Tracealyzer
+
+Zephyr offers support to third party tools that use the tracing subsystem and extend some of its functionalities in order to provide more resources. One of them is Percepio's Tracealyzer. To enable it, the following configs must be set:
+
+```text
+CONFIG_TRACING=y
+CONFIG_PERCEPIO_TRACERECORDER=y
+```
+
+The directions to configure a streaming port or use a RAM buffer to store the logged data can be found in reference [[4]].
+
+Tracealyzer offers a front-end tool to analyze and process the collected data, helping visualize how the system is behaving dynamically:
+
+<a href="img/img004.webp" target="_blank">
+    {{< figure
+        default=true
+        src="img/img004.webp"
+        alt=""
+        caption="Tracealyzer GUI"
+    >}}
+</a>
+
+The process of instrumenting the code and selecting the right data to be acquired is analogous to using an oscilloscope or other kind of probe to evaluate hardware. One might need to place additional *hooks* (instrumentation for the tracing system to acquire data at a specific point) in the application code and change the base configuration to select or limit data acquisition to interest.

```

Comment:
`to interest` is kind of hanging in the air here. How about something like this:

```suggestion
The process of instrumenting the code and selecting the right data to be acquired is analogous to using an oscilloscope or other kind of probe to evaluate hardware. One might need to place additional *hooks* (instrumentation for the tracing system to acquire data at a specific point) in the application code and change the base configuration to narrow data acquisition down to the areas of interest.

```